### PR TITLE
WebApp update - Code quality

### PR DESCRIPTION
**Pull request by Omarley7**
- Many issues from code climate have been addressed.
- Added .gittattibute so WSL can run scripts on Windows machines with configuring line endings locally
- Readme has been updated
- React-tabs have been implemented instead of MUI tabs to reduce code and accommodate code climate issues.
- Extended E2E test on menu to cover 100% of navigation

### CODE REVIEW

File: client/test/e2e/Menu.test.ts [hunk 1]
```diff
@@ -2,100 +2,52 @@
...
 
-  test('Navigation Links on the Scenario Analysis page', async ({ page }) => {
-    await page.goto('/sanalysis');
-    await page.locator('div[role="button"]:has-text("Dashboard")').click();
-    await expect(page).toHaveURL('/dashboard');
-
-    await page.goto('/sanalysis');
-    await page.locator('div[role="button"]:has-text("Library")').click();
-    await expect(page).toHaveURL('/library');
+  links.forEach((link) => {
```
Comment (prasadtalasila) [inline client/test/e2e/Menu.test.ts:34]: This is a strange convention. Usually we put code inside test.describe() function. Now we are iterating on the test() function itself. Can you move the link.forEach() into test.describe()?

File: client/test/e2e/Menu.test.ts [hunk 2]
```diff
@@ -2,100 +2,52 @@
...
-    await expect(page).toHaveURL('/dashboard');
-
-    await page.goto('/library');
-    await page.locator('div[role="button"]:has-text("Library")').click();
-    await expect(page).toHaveURL('/library');
-
-    await page.goto('/library');
-    await page.locator('div[role="button"]:has-text("Digital Twins")').click();
-    await expect(page).toHaveURL('/digitaltwins');
+interface TestLink {
```
Comment (prasadtalasila) [inline client/test/e2e/Menu.test.ts:5]: nice refactoring

File: client/src/components/tab/TabComponent.tsx [hunk 3]
```diff
@@ -1,82 +1,37 @@
 import * as React from 'react';
-import Tabs from '@mui/material/Tabs';
-import Tab from '@mui/material/Tab';
-import Box from '@mui/material/Box';
-import Paper from '@mui/material/Paper';
-import TabPanel, { a11yProps } from 'components/tab/subcomponents/TabPanel';
-import { useState, SyntheticEvent } from 'react';
+import { Paper } from '@mui/material';
+import TabRender from './subcomponents/TabRender';
```
Comment (prasadtalasila) [inline client/src/components/tab/TabComponent.tsx:3]: There seems to be a circular dependency between TabComponents and TabRender.

Comment (prasadtalasila) [reply client/src/components/tab/TabComponent.tsx:3]: Also as a convention, things in parent directories depend on sub-directories, except when they are in different packages, or in different parts of directory structure. For example, code in src/components/subcomponents/TabRender.tsx can depend on src/utility/sample.tsx but not on src/components/Components.tsx.

**Review by prasadtalasila — CHANGES_REQUESTED** [review]
Other comments: I've missed these changes over time. The react-iframe npm package has been removed from src/components/Iframe.tsx. Doesn't it provide better functionality than the existing code?

**Comment by prasadtalasila** [comment]
At this moment, about 75% of the code quality issues are in client/. Is it possible to fix more code quality issues in this PR?

File: client/src/components/tab/TabComponent.tsx (hunk 3 above)
Comment (Omarley7) [reply client/src/components/tab/TabComponent.tsx]: @DaloonOfDoom took a look at this and fixed it

File: client/test/unitTests/__mocks__/page_mocks.tsx [hunk 4]
```diff
@@ -0,0 +1,22 @@
+import * as React from 'react';
+
+jest.mock('page/Layout', () => ({
+  default: ({ children }: { children: React.ReactNode }) => (
```
[the remark this answers is not in the document]
Comment (Omarley7) [reply client/test/unitTests/__mocks__/page_mocks.tsx:4]: The other location is placeholder code. This will not be fixed.

**Comment by prasadtalasila** [comment]
@Omarley7, the scope of this PR seems to be expanding quite a bit. Can you please list the improvements and downgrades made in this PR so far?
If there are already significant improvements, let's merge this and push further work into next PR. Thanks.

**Comment by Omarley7** [comment]
@prasadtalasila I was hoping to include the issue discussed in #42.
I'll make an update on the work in this PR and move some of the E2E test stuff to the next PR.

**Comment by Omarley7** [comment]
Here's the update @prasadtalasila 

## Scope of this PR
Priority number 1 was to reduce CodeClimte issues. To achieve that the following changes have been made:
- React-tabs have been implemented instead of MUI tabs to reduce code and complexity. It uses styled-components paradigm
- All unit tests are now using global mocks and test utility. Greatly improves the implementation of the testing framework. Described in detail [here](https://github.com/Omarley7/DTaaS-Bachelor-new-GUI/pull/48#issue-1673906384).
- Bash scripts now declare PATH before exporting
- Removed middle routes like "workflows" and "components", as they are not needed yet and only increase the complexity 

### Pipeline improvements
- Added .gittattibute so WSL can run scripts on Windows machines with configuring line endings locally
- `.codeclimate.yml` has been updated from V1 to V2. Runs much better locally. It uses .eslintrc from the client since it was the only way we could make it work. You mentioned this as well in #24. Any reason why it was ruled out?
- We have removed husky from the project, as git-hooks provide the same functionality and take care of #34 It has also been added to the `<root>/README.md`

### Client improvements
- `.eslintrc` has been corrected according to #36 
- Readme has been updated to improve installation processes (could use some feedback on this from @KarstenMalle or @Artin13. Does it make sense when you read it?)
- React-Iframe has been reinstated. It was removed early on because its value wasn't understood in the beginning. Now we realize the type-safety is much better with this package.
- Added interface description to data files in `route` directories
- On the Digital Twin route, all iframe tabs now go to URL/label: e.g. `https://example.com/models`
- Instead of using environment variables directly in the code, getter functions have been implemented in `util/envUtils.ts` Greatly improves testability.


## Not covered yet
- E2E test still produces codeclimate issues. Should be fairly easy to fix in the next PR.
- Pipeline (`client.yml`) is not catching test errors #42

File: client/test/e2e/Menu.test.ts (hunk 1 above)
Comment (Omarley7) [reply client/test/e2e/Menu.test.ts]: I'm pretty sure I read somewhere that it's also convention to loop outside of the test() function if needed (of course I failed to find where I read it). 

- In this case, we will have identical tests for each route. (Loop 1)
- Each route navigates to every link to check if it works. (Loop 2)

Looping twice inside the test gave us many issues.

For now I've taken it out of the PR, and will look into it more in detail for the next PR.

**Comment by KarstenMalle** [comment]
@Omarley7 

Readme looks good to us, only thing is that we have to use sudo for the bash script/install, to install correctly since we use a Ubuntu VM, but i'm guessing it's not required with the suggested setup of "vagrant virtual machine".

**Pull request merged**